File: arcgis_map_sdk_android/android/src/main/kotlin/dev/fluttercommunity/arcgis_map_sdk_android/ManualLocationDataSource.kt
package dev.fluttercommunity.arcgis_map_sdk_android

import com.esri.arcgisruntime.location.LocationDataSource
import dev.fluttercommunity.arcgis_map_sdk_android.model.UserPosition
import dev.fluttercommunity.arcgis_map_sdk_android.model.toAGSPoint

class ManualLocationDisplayDataSource : LocationDataSource() {

    override fun onStart() {
        this.onStartCompleted(null)
    }

    override fun onStop() {

    }

    fun setNewLocation(userPosition: UserPosition) {
        val loc = Location(
            userPosition.latLng.toAGSPoint(),
            userPosition.accuracy ?: 0.0,
            userPosition.velocity ?: 0.0,
            userPosition.heading ?: 0.0,
            false
        )
        updateLocation(loc)
    }
}

File: arcgis_map_sdk_ios/ios/arcgis_map_sdk_ios/Sources/arcgis_map_sdk_ios/ManualLocationDataSource.swift
//
//  ManualLocationDataSource.swift
//  arcgis_map_sdk_ios
//
//  Created by Julian Bissekkou on 27.11.23.
//

import Foundation
import ArcGIS

final class CustomLocationProvider: LocationProvider {
    private var locationContinuation: AsyncThrowingStream<Location, Error>.Continuation?
    private var headingContinuation: AsyncThrowingStream<Double, Error>.Continuation?
    
    
    // Exposed stream
    var locations: AsyncThrowingStream<Location, Error> {
        AsyncThrowingStream { continuation in
            self.locationContinuation = continuation
        }
    }
    
    var headings: AsyncThrowingStream<Double, Error> {
        AsyncThrowingStream { continuation in
            self.headingContinuation = continuation
        }
    }

    // Push location from outside
    public func setNewLocation(_ position: UserPosition) {
        let loc = Location(
            position: position.latLng.toAGSPoint(),
            horizontalAccuracy: position.accuracy ?? 0,
            verticalAccuracy: position.accuracy ?? 0,
            speed: position.velocity ?? 0,
            course: position.heading ?? 0,
            isLastKnown: false
        )
        locationContinuation?.yield(loc)
        if let heading = position.heading {
            headingContinuation?.yield(heading)
        }
    }
}

